### PR TITLE
Add commit API endpoint

### PR DESCRIPTION
**Pull request by dinooo13**
Hi @clue this is my first try at a contribution, I hope it's any good.

### CODE REVIEW

**Comment by dinooo13** [comment]
friendly reminder 😄

File: src/Client.php [hunk 1]
```diff
@@ -1186,6 +1186,39 @@ public function imageSearch($term)
...
+     * @param string $repo      Repository name for the created image
+     * @param string $tag       Tag name for the created image
+     * @param string $comment   Commit message
+     * @param string $author    Author of the image (e.g., John Hannibal Smith <hannibal@a-team.com>)
+     * @param bool   $pause     Whether to pause the container before committing, default: true
+     * @param string $changes   Dockerfile instructions to apply while committing
+     * @return PromiseInterface
+     * @link https://docs.docker.com/engine/api/v1.41/#operation/ImageCommit
+     */
+    public function commit($config, $container, $repo = null, $tag = null, $comment = null, $author = null, $pause = true, $changes = null)
```
Comment (clue) [inline src/Client.php:1203]: Are you aware of any other API implementations of this interface? I wonder if it makes sense to perhaps name this `containerCommit()` to be more in line with the existing API methods?

Also, are you aware of any other implementations that take advantage of the `$config` array? I wonder if we should prominently feature the `$config`, use the `$container` as first argument or even drop a number of these parameters?

File: tests/ClientTest.php [hunk 2]
```diff
@@ -601,6 +601,15 @@ public function testImageSearch()
         $this->expectPromiseResolveWith($json, $this->client->imageSearch('clue'));
     }
 
+    public function testCommit()
+    {
+        $json = array();
+        $config = array();
+        $this->expectRequestFlow('post', 'commit?container=123&pause=1', $this->createResponseJson($json), 'expectJson');
+
+        $this->expectPromiseResolveWith($json, $this->client->commit($config, '123'));
```
Comment (clue) [inline tests/ClientTest.php:610]: LGTM :+1: Can you add another test that checks this with all parameters set?

**Review by clue — COMMENTED** [review]
@dinooo13 Thank you for looking into this, very nice contribution! Agree that this feature makes perfect sense! (closes #16)

I've added some minor remarks below, otherwise LGTM. Can you squash all your changes once done?

Thank you again, keep it up! :+1:

File: tests/ClientTest.php (hunk 2 above)
Comment (dinooo13) [reply tests/ClientTest.php]: Added!

File: src/Client.php (hunk 1 above)
Comment (SimonFrings) [reply src/Client.php]: Hey @dinooo13, thanks for your work on this so far :+1: 

@clue and I tried this out using docker one the command line and we looked at the necessary parameters when committing a container. It seems like the container parameter is the most important one in there followed by the repository and tag parameters (nearly the same as your suggestion).
We didn't find a config parameter tho, so I'm wondering what your use case for this is, to get an better understanding on why you added the `$config` parameter.

Comment (dinooo13) [reply src/Client.php]: Hey @SimonFrings I found this in the Docke Engine API docs.
https://docs.docker.com/engine/api/v1.41/#operation/ImageCommit
I'm not a Docker expert by any means, so I just oriented myself to it, maybe it's for configuration on top of the used container. But I guess YAGNI and should I update the PR to be more inline with the command line order of params?

Comment (SimonFrings) [reply src/Client.php]: You're right, they implement it too, the only difference is that they put the `$config` parameter at the end of the parameter list. I'm not the one too judge if this is needed or not so I'm okay with adding this as the last parameter.

What's your opinion on this?

Comment (dinooo13) [reply src/Client.php]: I will update the PR so it will be the last parameter. Thats a good solution imo.

Comment (dinooo13) [reply src/Client.php]: PR is updated 😄

Comment (dinooo13) [reply src/Client.php]: It's squashed :shipit:

**Review by SimonFrings — APPROVED** [review]
Nice work 👍

File: src/Client.php [hunk 3]
```diff
@@ -1186,6 +1186,39 @@ public function imageSearch($term)
...
+     * @param string $tag       Tag name for the created image
+     * @param string $comment   Commit message
+     * @param string $author    Author of the image (e.g., John Hannibal Smith <hannibal@a-team.com>)
+     * @param bool   $pause     Whether to pause the container before committing, default: true
+     * @param string $changes   Dockerfile's instructions to apply while committing
+     * @param array  $config    The container configuration (body of the request)
+     * @return PromiseInterface
+     * @link https://docs.docker.com/engine/api/v1.41/#operation/ImageCommit
+     */
+    public function containerCommit($container, $repo = null, $tag = null, $comment = null, $author = null, $pause = true, $changes = null, $config = array())
```
Comment (clue) [inline src/Client.php:1203]: * Docblocks should mention `null` values like `@param ?string $repo` etc. (see also other params)
* Last parameter type should be `array $config = array()` (other types would require PHP 7.1+)

**Review by clue — COMMENTED** [review]
@dinooo13 Thank you for looking into this and providing this high-quality PR, much appreciated! :+1: 

The `containerCommit()` API name is *interesting*, given the API docs use ["ImageCommit"](https://docs.docker.com/engine/api/v1.41/#operation/ImageCommit) ("Create a new image from a container") and the API endpoint and docker CLI command are both named just "commit". What do you think about this?

Also noticed two minor issues below, can you look into this?

Keep it up! :+1:

File: src/Client.php (hunk 3 above)
Comment (dinooo13) [reply src/Client.php]: Added nullable to the docblock. But I don't quite understand the second point should'nt the type be omitted for backwards compatibility?

Comment (clue) [reply src/Client.php]: Full type support would indeed require PHP 7.1+, but the `array` type declaration is already supported in older PHP versions: https://3v4l.org/12MCi

Comment (dinooo13) [reply src/Client.php]: Oh okay I didn't know that thanks I will add it to the PR 👍🏼

Comment (dinooo13) [reply src/Client.php]: Updated

**Review by clue — APPROVED** [review]
@dinooo13 Thank you very much for the update, changes LGTM! :shipit: Keep it up! :+1:

**Pull request merged**